design/FY2021/NFP-Waterside-Economizer.md

```diff
@@ -0,0 +1,632 @@
+
+            } else if (SELECT_CASE_var == CoilUserDefined) { // Coil:UserDefined
+                SimCoilUserDefined(state, CompName, CompIndex, AirLoopNum,
+                                   HeatingActive, CoolingActive);
+
+            } else if (SELECT_CASE_var == UnitarySystemModel) {
+                // 'AirLoopHVAC:UnitarySystem'
+                Real64 sensOut = 0.0;
+                Real64 latOut = 0.0;
+                CompPointer->simulate(state,
+                                      CompName,
+                                      FirstHVACIteration,
+                                      AirLoopNum,
+                                      CompIndex,
+                                      HeatingActive,
+                                      CoolingActive,
+                                      OAUnitNum,
+                                      OAUCoilOutTemp,
+                                      ZoneEquipFlag,
+                                      sensOut,
+                                      latOut);
+
+            **New simulate call for CoilWaterSystem using override for HVACSystemData::simulate:**
+
+            else if (SELECT_CASE_var == CoilSystemWater) { // 'CoilSystem:Cooling:Water'
+			
+				if (CompPointer == nullptr) {
+					UnitarySystems::UnitarySys thisSys;
+					CompPointer = thisSys.factory(state, DataHVACGlobals::UnitarySys_AnyCoilType, CompName, false, 0);
+					// temporary fix for saving pointer, eventually apply to UnitarySystem 25 lines down
+					state.dataAirSystemsData->PrimaryAirSystems(airLoopNum).Branch(branchNum).Comp(compNum).compPointer = CompPointer;
+				}			
+                Real64 sensOut = 0.0;
+                Real64 latOut = 0.0;
+                CompPointer->simulate(state,
+                                      CompName,
+                                      FirstHVACIteration,
+                                      AirLoopNum,
+                                      CompIndex,
+                                      HeatingActive,
+                                      CoolingActive,
+                                      OAUnitNum,
+                                      OAUCoilOutTemp,
+                                      ZoneEquipFlag,
+                                      sensOut,
+                                      latOut);
+
+
+Add component type for 'CoilSystem:Cooling:Water' in Primary Air Loop
+    constexpr int Fan_System_Object(28);
+    constexpr int UnitarySystemModel(29);
+    constexpr int ZoneVRFasAirLoopEquip(30);
+
+    Add component type here
+    constexpr int CoilWaterSystems(31);
+
+
+### DataPlant.cc ###
+    Add "CoilSystem:Cooling:Water" system cooling coil to the 1D String lists:
+
+	SimPlantEquipTypes
+	ccSimPlantEquipTypes
+	ValidLoopEquipTypes
+
+	Change NumSimPlantEquipTypes(96) to NumSimPlantEquipTypes(97);
+
+	Add parameter index for "CoilSystem:Cooling:Water" component
+    constexpr int TypeOf_CoilSystemCoolingWater(97);
```

Comment:
I don't think these plant changes are necessary. The coil object is what will go directly on the plant Branch. The CoilSystem is just a wrapper that shows on an airloop branch or OA system list.

Reply:
You are right. I was thinking it might need the change but I did not need it to run a test file in working branch. Lesson learned. @mjwitte Thanks.

Reply:
@mjwitte Now I am running using a new "getInput" function to read in "COILSYSTEM:COOLING:WATER" and instantiate the object I read as UnitarySystem. It runs fine and I get almost identical results (diffs within tolerance limits) as just running native Unitary System. But I get the following warning error message:

```
   ** Warning ** Potential Node Connection Error for object COILSYSTEM:COOLING:WATER, name=UNITARY-FREE-COOLING
   **   ~~~   **   Node Types are still UNDEFINED -- See Branch/Node Details file for further information
   **   ~~~   **   Inlet Node : NODE 102
   **   ~~~   **   Outlet Node: NODE 37
   ** Warning ** Node Connection Error for object AIRLOOPHVAC:UNITARYSYSTEM, name=UNITARY-FREE-COOLING
   **   ~~~   **   Air Nodes not on any Branch or Parent Object
   **   ~~~   **   Inlet Node : NODE 102
   **   ~~~   **   Outlet Node: NODE 37
```

This is expected error for now but since Yesterday I am getting empty error file (same as issue #8788 you posted) and I am not sure what is causing it. I tested develop by switching to develop in current working branch after cleaning the build and had the same problem. But when I tested develop with new clone project the issue went away. I am puzzled what is causing it.

Reply:
My guess is that the airloop branch object in the idf says "AIRLOOPHVAC:UNITARYSYSTEM" instead of "COILSYSTEM:COOLING:WATER"? Either that, or one of the `*CompSet` calls either in UnitarySystem or in the Branch is using the wrong equipment type.

Regarding the empty err file, I'm not sure. Is your test file using PixelCounting? If not, perhaps this problem is not directly related to that?

Reply:
@mwetter OK. Resolved now. It appears that the one code change that I made yesterday resulted in these issues. I removed initialization of the a pointer to the new object during getinput, instead I set to initialize the pointer only during the first call for simulation. Apparently, it didn't like it. Now reverted that change, the test file runs fine and the error output file is not empty.  So, it appears the problem is triggered by a pointer initialization problem or late pointer initialization.

Reply:
It is not using PixelCounting? Both problems are gone.

Reply:
I was thinking that we read the object "COILSYSTEM:COOLING:WATER" in the air loop branch but then we initialize the object at get input as "AIRLOOPHVAC:UNITARYSYSTEM" type, so we may have different component type info and believed that the data in different part are not in sink. I do not see that problem now, may be one of the changes I did yesterday must have resolved it, I think.

Reply:
My expectation is that it will remain tagged at "COILSYSTEM:COOLING:WATER" even though UnitarySystem code is processing it. If you check the bnd output file, COILSYSTEM:COOLING:WATER should be there for the various node connection blocks.

Reply:
I had a similar Node Connection issue in 8729 with OutdoorAirUnit when UnitarySystem was a child. Not sure this is the same thing but it was a timing issue with SetUpCompSets. Child needs to call SetUpCompSets before parent?

Reply:
@This issue is gone now. It must have been a timing issue.